apps/rush-lib/src/logic/pnpm/PnpmShrinkwrapFile.ts

```diff
@@ -505,75 +506,82 @@ export class PnpmShrinkwrapFile extends BaseShrinkwrapFile {
+        // See: https://github.com/pnpm/pnpm/blob/main/packages/lockfile-utils/src/satisfiesPackageManifest.ts
+        switch (foundDependency.dependencyType) {
+          case DependencyType.Optional:
+            break;
+          case DependencyType.Regular:
+            if (packageDependency.dependencyType === DependencyType.Optional) {
+              dependencyVersions.set(packageDependency.name, packageDependency);
+            }
+            break;
+          case DependencyType.Dev:
+            dependencyVersions.set(packageDependency.name, packageDependency);
+            break;
+        }
+      }
     }
 
-    const dependencyKey: string = allDependencies[packageName];
-    return this._parsePnpmDependencyKey(packageName, dependencyKey);
-  }
-
-  /**
-   * Returns the version of a dependency being used by a given project
-   */
-  private _getDependencyVersion(
-    dependencyName: string,
-    tempProjectName: string
-  ): DependencySpecifier | undefined {
-    const tempProjectDependencyKey: string | undefined = this.getTempProjectDependencyKey(tempProjectName);
-    if (!tempProjectDependencyKey) {
-      throw new Error(`Cannot get dependency key for temp project: ${tempProjectName}`);
+    // Then validate that the dependency fields are as expected in the shrinkwrap to avoid false-negatives
+    // when moving a package from one field to the other.
+    for (const dependencyVersion of dependencyVersions.values()) {
+      switch (dependencyVersion.dependencyType) {
+        case DependencyType.Optional:
+          if (!importer.optionalDependencies[dependencyVersion.name]) return true;
+          break;
+        case DependencyType.Regular:
+          if (!importer.dependencies[dependencyVersion.name]) return true;
+          break;
+        case DependencyType.Dev:
+          if (!importer.devDependencies[dependencyVersion.name]) return true;
+          break;
+      }
     }
 
-    const packageDescription: IPnpmShrinkwrapDependencyYaml | undefined = this._getPackageDescription(
-      tempProjectDependencyKey
-    );
-    if (!packageDescription || !packageDescription.dependencies) {
-      return undefined;
+    // Then validate the length matches between the importer and the dependency list, since duplicates are
+    // a valid use-case. Importers will only take one of these values, so no need to do more work here.
+    if (dependencyVersions.size !== Object.keys(importer.specifiers).length) {
+      return true;
     }
 
-    if (!packageDescription.dependencies.hasOwnProperty(dependencyName)) {
-      return undefined;
+    // Finally, validate that all values in the importer are also present in the dependency list.
+    for (const [importerPackageName, importerVersionSpecifier] of Object.entries(importer.specifiers)) {
+      const foundDependency: PackageJsonDependency | undefined = dependencyVersions.get(importerPackageName);
+      if (!foundDependency || foundDependency.version !== importerVersionSpecifier) {
```

Comment:
We might want to run the package.json through pnpmfile.js before this test. Otherwise we give people confusing results if their pnpmfile overrides a local package's dependencies (i.e. pnpm is happy, rush is not)

Reply:
We already check to see if pnpmfile has been modified and prevent skipping installs if the file is updated. We also officially document that running `rush update --full` after updating pnpmfile.js is recommended in the template pnpmfile that gets dropped after running `rush init`, as well as on the rushjs.io website: https://rushjs.io/pages/configs/pnpmfile_js/

I'm going to leave this as-is, since the worst that can happen is someone runs into this case and has to run `rush update` now that the packages are changed on-disk. PNPM will still block install with an issue saying that the shrinkwrap is out of date, and update will still be able to be run.